syftlib/src/main/java/org/openmined/syft/SignallingClient.kt

```diff
@@ -0,0 +1,141 @@
+    private val responseObservable: Flowable<String>
+    private val keepSendingScheduler = Executors.newScheduledThreadPool(1)
+    private var sendingThread: Future<*>
+
+    init {
+        responseObservable = Flowable.create<String>(
+            { emitter ->
+                emitter.setCancellable { this.stop() }
+                try {
+                    createSocket()
+                    while (socket.isConnected) {
+                        emitter.onNext(reader.readLine())
+                    }
+                    emitter.onComplete()
+                } catch (e: Exception) {
+                    Log.e(TAG, "Socket TimeOut", e)
+                    stop()
+                    emitter.onError(e)
+                }
+            }, BackpressureStrategy.BUFFER
+        )
+
+        disposable = invokeSocketObserver()
+        sendingThread = runKeepAliveSchedule()
+    }
+
+    /**
+     * Send the data over the Socket connection to PyGrid
+     */
+    fun send(type: String, data: JsonObject) {
+        if (socket.isConnected) {
+            sendingThread.cancel(false)
+            data.addProperty("workerId", workerId)
+            val message = JsonObject().apply {
+                this.addProperty("type", type)
+                this.add("data", data)
+            }
+            writer.write(message.toString())
+            writer.flush()
+            Log.d(TAG, "message sent :$message")
+            sendingThread = runKeepAliveSchedule()
+        } else {
+            disposable.dispose()
+            invokeSocketObserver()
+        }
+
+    }
+
+    private fun runKeepAliveSchedule(): ScheduledFuture<*> {
+        return keepSendingScheduler.scheduleAtFixedRate(
+            {
+                if (socket.isConnected)
+                    send(SOCKET_PING, JsonObject())
+            },
+            keepAliveTimeout.toLong(),
+            keepAliveTimeout.toLong(),
+            TimeUnit.MILLISECONDS
+        )
+    }
+
+    /**
+     * @return void but sets up SignallingClient class properties
+     */
+    private fun createSocket() {
+        Log.d(TAG, "Opening Socket Connection")
+        socket = Socket(url, port)
+        Log.d(TAG, "Socket Connected")
+        socket.keepAlive = true
+        socket.soTimeout = keepAliveTimeout
+        writer = socket.getOutputStream().bufferedWriter(Charset.defaultCharset())
+        reader = socket.getInputStream().bufferedReader(Charset.defaultCharset())
+        signallingInterface.onOpen()
+    }
+
+    /**
+     * Whenever a subscriber is subscribed to Observable, a new invocation happens
+     * initiating socket connection and setting up socket reader and writer
+     *
+     * This function is also called when connection is no more `Connected`
+     * thereby re-establishing connection to PyGrid
+     */
+    private fun invokeSocketObserver(): Disposable {
+        return responseObservable
+            .observeOn(AndroidSchedulers.mainThread())
```

Comment:
It would be better if the schedulers are provided to this class. This way will allow the app using this class to determine where results are going to be processed.